fdbserver/TagPartitionedLogSystem.actor.cpp

```diff
@@ -2080,41 +2086,135 @@ ACTOR Future<Void> TagPartitionedLogSystem::getDurableVersionChanged(LogLockInfo
+		std::unordered_map<Version, boost::dynamic_bitset<>> versionAvailableTLogs;
+		// version -> all tLogs that the version was sent to (by the commit proxy)
+		std::map<Version, boost::dynamic_bitset<>> versionAllTLogs;
+		// version -> prevVersion (that was given out by the sequencer) map
+		std::map<Version, Version> prevVersionMap;
 		int replicationFactor = std::get<0>(logGroupResult);
 		for (auto& tLogResult : std::get<1>(logGroupResult)) {
+			uint16_t tLogLocId = tLogLocIds[tLogGroupIdx][tLogIdx];
+			availableTLogs.set(tLogLocId);
 			bool logGroupCandidate = false;
 			for (auto& unknownCommittedVersion : tLogResult.unknownCommittedVersions) {
 				Version k = std::get<0>(unknownCommittedVersion);
 				if (k > minEnd) {
-					versionRepCount[k]++;
-					versionTLogCount[k] = std::get<1>(unknownCommittedVersion);
+					if (versionAvailableTLogs[k].empty()) {
+						versionAvailableTLogs[k].resize(bsSize);
+					}
+					versionAvailableTLogs[k].set(tLogLocId);
+					prevVersionMap[k] = std::get<1>(unknownCommittedVersion);
+					if (versionAllTLogs[k].empty()) {
+						versionAllTLogs[k].resize(bsSize);
+					}
+					populateBitset(versionAllTLogs[k], std::get<2>(unknownCommittedVersion));
 					logGroupCandidate = true;
 				}
 			}
 			if (!logGroupCandidate) {
 				return minEnd;
 			}
+			tLogIdx++;
 		}
 		Version minTLogs = minEnd;
-		for (auto const& [version, tLogCount] : versionTLogCount) {
-			if (versionRepCount[version] >= tLogCount - replicationFactor + 1) {
-				minTLogs = version;
-			} else {
+		Version prevVersion = minEnd;
+		for (auto const& [version, tLogs] : versionAllTLogs) {
+			if (!(prevVersion == minEnd || prevVersion == prevVersionMap[version])) {
 				break;
 			}
+			// This version is not recoverable if there is a log server (LS) such that:
+			// - the commit proxy sent this version to LS (i.e., LS is present in "versionAllTLogs[version]")
+			// - LS is available (i.e., LS is present in "availableTLogs")
+			// - LS didn't receive this version (i.e., LS is not present in "versionAvailableTLogs[version]")
+			if (((tLogs & availableTLogs) & ~versionAvailableTLogs[version]).any()) {
+				break;
+			}
+			// If the commit proxy sent this version to "N" log servers then at least
+			// (N - replicationFactor + 1) log servers must be available.
+			if (!(versionAvailableTLogs[version].size() >= tLogs.size() - replicationFactor + 1)) {
+				break;
+			}
+			// Update RV.
+			minTLogs = version;
+			// Update recovery version vector.
+			for (boost::dynamic_bitset<>::size_type id = 0; id < versionAvailableTLogs[version].size(); id++) {
+				if (versionAvailableTLogs[version][id]) {
+					RVs[id] = version;
+				}
+			}
+			// Update prevVersion.
+			prevVersion = version;
 		}
 		minLogGroup = std::min(minLogGroup, minTLogs);
+		tLogGroupIdx++;
 	}
+	ASSERT_WE_THINK(minLogGroup >= minDVEnd && minLogGroup != std::numeric_limits<Version>::max());
```

Comment:
should this be `ASSERT` rather than test-only `ASSERT_WE_THINK` ?

Reply:
Let's keep this for now and see what we find in simulation tests. It is likely this assertion is true in which case we can start from "min(DV)" itself, and we won't even need this assert.